infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/LocalCheInfrastructureProvisioner.java

```diff
@@ -82,56 +88,93 @@ public LocalCheInfrastructureProvisioner(InstallerConfigApplier installerConfigA
     }
 
     @Override
-    public void provision(Environment envConfig,
+    public void provision(EnvironmentImpl envConfig,
                           DockerEnvironment internalEnv,
                           RuntimeIdentity identity)
             throws InfrastructureException {
         String devMachineName = getDevMachineName(envConfig);
         if (devMachineName == null) {
             throw new InfrastructureException("ws-machine is not found on installers applying");
         }
+        excludeDirsFromSnapshot(internalEnv);
+        addProjects(internalEnv, identity.getWorkspaceId(), devMachineName);
+        addInstallers(envConfig, internalEnv);
+        addLabels(envConfig, internalEnv, identity);
+    }
 
-        DockerContainerConfig devMachine = internalEnv.getContainers().get(devMachineName);
+    private void addInstallers(EnvironmentImpl envConfig, DockerEnvironment internalEnv)
+            throws InfrastructureException {
 
-        for (DockerContainerConfig machine : internalEnv.getContainers().values()) {
-            ArrayList<String> volumes = new ArrayList<>(machine.getVolumes());
-            volumes.add(terminalVolumeProvider.get());
-            volumes.add(execVolumeProvider.get());
-            machine.setVolumes(volumes);
-        }
+        for (Map.Entry<String, MachineConfigImpl> machineConfigEntry : envConfig.getMachines().entrySet()) {
+            String machineName = machineConfigEntry.getKey();
+            MachineConfigImpl machineConfig = machineConfigEntry.getValue();
+            DockerContainerConfig containerConfig = internalEnv.getContainers().get(machineName);
 
-        // add bind-mount volume for projects in a workspace
-        String projectFolderVolume;
-        try {
-            projectFolderVolume = String.format("%s:%s%s",
-                                                workspaceFolderPathProvider.getPath(identity.getWorkspaceId()),
-                                                projectFolderPath, projectsVolumeOptions);
-        } catch (IOException e) {
-            throw new InfrastructureException("Error occurred on resolving path to files of workspace " +
-                                              identity.getWorkspaceId());
+            for (String installer : machineConfig.getInstallers()) {
+                switch (installer) {
+                    case "org.eclipse.che.terminal":
```

Comment:
I've seen a few places where installer identifiers used as string literals, 
what if we create constants for them?